sql/index.go

```diff
@@ -87,18 +99,272 @@ type Mergeable interface {
+	Create(path, db, id string, expr Expression) (Index, error)
 	// Load the index at the given path.
 	Load(path string) (Index, error)
+	// Save the given index at the given path.
+	Save(ctx context.Context, path string, index Index, iter IndexKeyValueIter) error
+	// Delete the index with the given path.
+	Delete(path string, index Index) error
 }
 
-// IndexSaver is the piece that stores indexes in disk.
-type IndexSaver interface {
-	// ID returns the unique name of the index saver.
-	ID() string
-	// Save the given index at the given path.
-	Save(index Index, path string) error
+type indexKey struct {
+	db, id string
+}
+
+// IndexRegistry keeps track of all indexes in the engine.
+type IndexRegistry struct {
+	root string
+
+	mut      sync.RWMutex
+	indexes  []Index
+	statuses map[indexKey]IndexStatus
+
+	driversMut sync.RWMutex
+	drivers    map[string]IndexDriver
+
+	rcmut            sync.RWMutex
+	refCounts        map[indexKey]int
+	deleteIndexQueue map[indexKey]chan<- struct{}
+}
+
+// NewIndexRegistry returns a new Index Registry.
+func NewIndexRegistry() *IndexRegistry {
+	return &IndexRegistry{
+		statuses:         make(map[indexKey]IndexStatus),
+		drivers:          make(map[string]IndexDriver),
+		refCounts:        make(map[indexKey]int),
+		deleteIndexQueue: make(map[indexKey]chan<- struct{}),
+	}
+}
+
+// IndexDriver returns the IndexDriver with the given ID.
+func (r *IndexRegistry) IndexDriver(id string) IndexDriver {
+	r.driversMut.RLock()
+	defer r.driversMut.RUnlock()
+	return r.drivers[id]
+}
+
+// RegisterIndexDriver registers a new index driver.
+func (r *IndexRegistry) RegisterIndexDriver(driver IndexDriver) {
+	r.driversMut.Lock()
+	defer r.driversMut.Unlock()
+	r.drivers[driver.ID()] = driver
+}
+
+func (r *IndexRegistry) retainIndex(db, id string) {
+	r.rcmut.Lock()
+	defer r.rcmut.Unlock()
+	key := indexKey{db, id}
+	r.refCounts[key] = r.refCounts[key] + 1
+}
+
+// CanUseIndex returns whether the given index is ready to use or not.
+func (r *IndexRegistry) CanUseIndex(idx Index) bool {
+	r.mut.RLock()
+	defer r.mut.RUnlock()
+	return bool(r.statuses[indexKey{idx.Database(), idx.ID()}])
+}
+
+// setStatus is not thread-safe, it should be guarded using mut.
+func (r *IndexRegistry) setStatus(idx Index, status IndexStatus) {
+	r.statuses[indexKey{idx.Database(), idx.ID()}] = status
+}
+
+// ReleaseIndex releases an index after it's been used.
+func (r *IndexRegistry) ReleaseIndex(idx Index) {
+	r.rcmut.Lock()
+	defer r.rcmut.Unlock()
+	key := indexKey{idx.Database(), idx.ID()}
+	r.refCounts[key] = r.refCounts[key] - 1
+
+	if r.refCounts[key] > 0 {
+		return
+	}
+
+	if ch, ok := r.deleteIndexQueue[key]; ok {
+		close(ch)
+		delete(r.deleteIndexQueue, key)
+	}
+}
+
+// Index returns the index with the given id. It may return nil if the index is
+// not found.
+func (r *IndexRegistry) Index(db, id string) Index {
+	r.mut.RLock()
+	defer r.mut.RUnlock()
+
+	id = strings.ToLower(id)
+	for _, idx := range r.indexes {
```

Comment:
maybe `indexes` can be a `map[db.id]Index` (if `id` is unique per db) so we can speed up a lookup

Reply:
We need to iterate anyway for the `IndexByExpression`, but ok, we can speed up getting by key.